File: services/http-gateway/app/controllers/api/v1/auth_controller.rb
# frozen_string_literal: true

module Api
  module V1
    class AuthController < ApplicationController
      def telegram
        user = TelegramLoginAuth.new.validate(auth_params)
        return render json: { error: "invalid telegram auth" }, status: :unauthorized unless user

        seller = CatalogServiceClient.new.get_seller_by_tg_id(user[:id])
        unless seller&.id&.positive?
          return render json: { error: "seller not found, run /start in bot first" }, status: :unauthorized
        end

        token = JwtSession.new.encode(seller_id: seller.id, tg_user_id: user[:id])
        cookies[JwtSession::COOKIE_NAME] = {
          value: token,
          httponly: true,
          secure: Rails.env.production?,
          same_site: :lax,
          path: "/"
        }

        render json: seller_json(seller)
      end

      def logout
        cookies.delete(JwtSession::COOKIE_NAME, path: "/")
        head :no_content
      end

      def me
        token = cookies[JwtSession::COOKIE_NAME]
        session = JwtSession.new.decode(token)
        return render json: { error: "unauthorized" }, status: :unauthorized unless session

        seller = CatalogServiceClient.new.get_seller(session[:seller_id])
        render json: seller_json(seller)
      rescue GRPC::NotFound
        render json: { error: "unauthorized" }, status: :unauthorized
      end

      private

      def auth_params
        params.permit(:id, :first_name, :last_name, :username, :photo_url, :auth_date, :hash)
      end

      def seller_json(seller)
        {
          id: seller.id,
          tg_user_id: seller.tg_user_id,
          username: seller.username,
          full_name: seller.full_name,
          sensitivity: seller.sensitivity,
          plan: seller.plan
        }
      end
    end
  end
end

File: services/core/src/main/kotlin/com/sellbot/core/http/AuthController.kt
package com.sellbot.core.http

import com.sellbot.core.http.auth.JWT_COOKIE_NAME
import com.sellbot.core.http.auth.JwtSession
import com.sellbot.core.http.auth.TelegramAuth
import com.sellbot.core.http.dto.toDto
import com.sellbot.core.service.CatalogService
import com.sellbot.core.service.TeamService
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.HttpHeaders
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseCookie
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RestController

@RestController
@RequestMapping("/api/v1/auth")
class AuthController(
    private val properties: SellbotHttpProperties,
    private val catalogService: CatalogService,
    private val teamService: TeamService,
    private val jwtSession: JwtSession,
) {

    @PostMapping("/telegram")
    fun telegramLogin(
        @RequestBody body: Map<String, Any?>,
        response: HttpServletResponse,
    ): ResponseEntity<Any> {
        val user = TelegramAuth.validateLoginWidget(body, properties.botToken)
            ?: return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(mapOf("error" to "invalid telegram auth"))

        val access = teamService.resolveDashboardAccess(user.id)
            ?: return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(mapOf("error" to "seller not found, run /start in bot first"))

        val token = jwtSession.encode(access.seller.id!!, user.id)
        setJwtCookie(response, token)
        return ResponseEntity.ok(access.seller.toDto(isOwner = access.isOwner))
    }

    @PostMapping("/logout")
    fun logout(response: HttpServletResponse): ResponseEntity<Void> {
        clearJwtCookie(response)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/me")
    fun me(request: jakarta.servlet.http.HttpServletRequest): ResponseEntity<Any> {
        return try {
            val seller = catalogService.getSeller(request.sellerId())
            val isOwner = teamService.isOwner(seller.id!!, request.tgUserId())
            ResponseEntity.ok(seller.toDto(isOwner = isOwner))
        } catch (_: Exception) {
            ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(mapOf("error" to "unauthorized"))
        }
    }

    private fun setJwtCookie(response: HttpServletResponse, token: String) {
        val cookie = ResponseCookie.from(JWT_COOKIE_NAME, token)
            .httpOnly(true)
            .secure(properties.secureCookies)
            .path("/")
            .sameSite("Lax")
            .maxAge(properties.jwt.ttlHours * 3600)
            .build()
        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString())
    }

    private fun clearJwtCookie(response: HttpServletResponse) {
        val cookie = ResponseCookie.from(JWT_COOKIE_NAME, "")
            .httpOnly(true)
            .secure(properties.secureCookies)
            .path("/")
            .sameSite("Lax")
            .maxAge(0)
            .build()
        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString())
    }
}
